### PR TITLE
Add GPU kernel for SparseFillEmptyRows[Grad]

### PR DESCRIPTION
**Pull request by benbarsdell**
Follow-up to https://github.com/tensorflow/tensorflow/pull/44719

cc @nluehr

### CODE REVIEW

**Comment by benbarsdell** [comment]
I fixed a few small issues I noticed since submitting this. I'm not sure why the Windows GPU build is complaining about std::complex on the device.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/sparse_fill_empty_rows_op.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op.cc]: Done.

File: tensorflow/python/kernel_tests/sparse_ops_test.py
Comment (benbarsdell) [reply tensorflow/python/kernel_tests/sparse_ops_test.py]: Done.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: Done.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:31]: Good idea, I'll do this after landing this PR.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:200]: The doc says `The output SparseTensor will be in row-major order` and I believe this behavior is relied-upon by some higher-level APIs.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: Done for the indices and values outputs. The other two require using `output_required()` which only accepts indexes, so I've left them as-is.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: Done.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: After looking at this I realized that we can report invalid indexes like the CPU does without adding much overhead, so I've done that and also added a test that the error gets raised.

**Comment by benbarsdell** [comment]
(Fixed Ubuntu CI issues).

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:204]: I believe "row major" in that context means that if the logical shape of the output is (say) `[A, B, C]` then the first element of an index corresponds to `A`, second one corresponds to `B` etc.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op.cc:198]: This could be a default value for `done` right?

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/core/kernels/sparse_fill_empty_rows_op.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op.cc:200]: I actually did that originally but the CI failed with some bizarre compiler/linker errors. I couldn't reproduce it myself but guessed that this was the problem. I've added a comment to that effect.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:204]: My understanding is that it is supposed to behave like tf.sparse.reorder, which is documented as "Reorders a SparseTensor into the canonical, row-major ordering."

There is a test for it: https://github.com/tensorflow/tensorflow/blob/4f80ad3/tensorflow/python/kernel_tests/sparse_ops_test.py#L588

The CPU implementation also explicitly checks if the rows are ordered, and only early-exits if they are: https://github.com/tensorflow/tensorflow/blob/69967c0/tensorflow/core/kernels/sparse_fill_empty_rows_op.cc#L129-L136

Comment (penpornk) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:204]: I agree with @benbarsdell. The behavior here is consistent with the CPU kernel:
* Can take 2D SparseTensors with unordered tuples.
* Returns a 2D SparseTensor with nonzeroes stored in row-major order.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op.cc
Comment (sanjoy) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op.cc:200]: SG

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:204]: SG, I'll proceed with the rest of the review.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:90]: So you don't need to compare columns because that's the "row major" invariant?  If so, please add a comment.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:80]: Is `rank` always `2`?

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:139]: Is this needed?  Won't we have returned an error if any indices were out of range?

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:402]: Can you please proactively change these to be `DCHECK`s to avoid accidentally holding up the merge?

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:450]: Can some of this logic be split into separate named functions?  I think that would make the flow more readable.

**Review by sanjoy — CHANGES_REQUESTED** [review]
One general comment: can you please refactor out some helper functions?  That will help readability I think.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:92]: That's correct, comment added.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:80]: No, it can be higher for higher-rank sparse tensors.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: Good point, removed.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc]: Done.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (benbarsdell) [reply tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:429]: I've pulled the output allocations and sorting logic out into helper methods so the lambda is a lot shorter now.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:359]: This routine (first get temp bytes, allocate temp byte and do actual work) can be a helper function.

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:336]: InclusiveSum

File: tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc
Comment (sanjoy) [inline tensorflow/core/kernels/sparse_fill_empty_rows_op_gpu.cu.cc:524]: Misplaced comment?

**Review by sanjoy — CHANGES_REQUESTED** [review]
Couple more cleanup suggestions.

**Comment by benbarsdell** [comment]
I've cleaned this up in https://github.com/tensorflow/tensorflow/pull/46172/commits/78353b59aa7f0b2bc88ad50f642a8044c1201468 using `gpu_prim_helpers.h` from the recently-merged https://github.com/tensorflow/tensorflow/pull/47082/. It addresses the three requested changes.

**Review by sanjoy — APPROVED** [review]

**Comment by sanjoy** [comment]
Ben, this PR looks functionally correct to me, but I believe these ops only work on sparse matrices, which means `rank` is always 2.  If you agree, do you mind sending a separate PR simplifying the kernel and adding the runtime check?

CC @penpornk @ebrevdo

**Comment by benbarsdell** [comment]
Thanks Sanjoy.

Regarding rank, the CPU implementation does currently support rank>2, and it is possible to pass a rank>2 `SparseTensor` to `sparse_fill_empty_rows` without getting any errors. The only indication I see of it being limited to rank=2 is this message in the documentation:`A SparseTensor with shape [N, M].`

Example with rank=3:
```
tf.sparse.fill_empty_rows(tf.sparse.SparseTensor(indices=[[0, 1, 2]], values=[3.14], dense_shape=[3, 4, 5]), default_value=1.23)
```

**Comment by sanjoy** [comment]
> Thanks Sanjoy.
> 
> Regarding rank, the CPU implementation does currently support rank>2, and it is possible to pass a rank>2 `SparseTensor` to `sparse_fill_empty_rows` without getting any errors. The only indication I see of it being limited to rank=2 is this message in the documentation:`A SparseTensor with shape [N, M].`

I suspect that's just a missing runtime check.  Otherwise the spec of the op (which talks in terms of "rows") does not make sense to me.

> 
> Example with rank=3:
> 
> ```
> tf.sparse.fill_empty_rows(tf.sparse.SparseTensor(indices=[[0, 1, 2]], values=[3.14], dense_shape=[3, 4, 5]), default_value=1.23)
> ```

**Comment by gbaned** [comment]
@benbarsdell  Can you please check @sanjoy's comments and keep us posted ? Thanks!

**Review by sanjoy — APPROVED** [review]

**Comment by benbarsdell** [comment]
Any info on why this was reverted?

**Pull request merged**
